src/util/assignment.rs

```diff
@@ -0,0 +1,256 @@
+use core::str;
+use std::{collections::HashMap, io::Read};
+
+use crypto_box::{
+    aead::{Aead, AeadCore, OsRng},
+    SalsaBox, PublicKey, SecretKey
+};
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use serde_with::serde_as;
+use serde_with::base64::Base64;
+use sha2::Sha512;
+use sha2::Digest;
+use sha3::digest::generic_array::GenericArray;
+use curve25519_dalek::edwards::CompressedEdwardsY;
+use flate2::read::GzDecoder;
+
+use crate::util::signature::timed_hmac_now;
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct Chunk {
+    pub id: String,
+    pub base_url: String,
+    pub files: HashMap<String, String>,
+    pub size_bytes: u64,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Dataset {
+    pub id: String,
+    pub base_url: String,
+    pub chunks: Vec<Chunk>,
+}
+
+#[derive(Debug, Default, Serialize, Deserialize)]
+#[serde(rename_all = "kebab-case")]
+struct Headers {
+    worker_id: String,
+    worker_signature: String,
+}
+
+#[serde_as]
+#[derive(Debug, Default, Serialize, Deserialize, Clone)]
+#[serde(rename_all = "camelCase")]
+struct EncryptedHeaders {
+    #[serde_as(as = "Base64")]
+    identity: Vec<u8>,
+    #[serde_as(as = "Base64")]
+    nonce: Vec<u8>,
+    #[serde_as(as = "Base64")]
+    ciphertext: Vec<u8>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct WorkerAssignment {
+    status: String,
+    chunks_deltas: Vec<u64>,
+    encrypted_headers: EncryptedHeaders,
+}
+
+#[derive(Debug, Default, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Assignment {
+    datasets: Vec<Dataset>,
+    worker_assignments: HashMap<String, WorkerAssignment>,
+    #[serde(skip)]
+    chunk_map: Option<HashMap<String, u64>>,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct NetworkAssignment {
+    pub(crate) url: String,
+    pub(crate) id: String,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct NetworkState {
+    pub(crate) network: String,
+    pub(crate) assignment: NetworkAssignment
+}
+
+impl Assignment {
```

Comment:
Can we move the common code to a crate inside sqd-network?

Reply:
Yep, it's next change as soon as it will be stabilised